cpp/distributed/dist-mnist.cpp

```diff
@@ -0,0 +1,201 @@
+        x = torch::relu(torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));
+        x = x.view({-1, 320});
+        x = torch::relu(fc1->forward(x));
+        x = torch::dropout(x, 0.5, is_training());
+        x = fc2->forward(x);
+        return torch::log_softmax(x, 1);
+    }
+
+    torch::nn::Conv2d conv1;
+    torch::nn::Conv2d conv2;
+    torch::nn::Dropout2d conv2_drop;
+    torch::nn::Linear fc1;
+    torch::nn::Linear fc2;
+};
+
+
+int main(int argc, char *argv[])
+{
+
+  //MPI variables
+  int rank, numranks;
+  MPI_Init(&argc, &argv);
+  MPI_Comm_size(MPI_COMM_WORLD, &numranks);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  //end MPI variables
+
+  // Timer variables
+  auto tstart = 0.0;
+  auto tend = 0.0;
+
+  //TRAINING
+  //Read train dataset
+  std::string filename = "/afs/crc.nd.edu/user/s/sghosh2/Public/ML/mnist/data";
+  auto train_dataset = torch::data::datasets::MNIST(filename)
+            .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
+            .map(torch::data::transforms::Stack<>());
+
+  //Distributed Random Sampler
+  auto data_sampler = torch::data::samplers::DistributedRandomSampler(train_dataset.size().value(),
+                                             numranks, rank, false);
+
+  auto num_train_samples_per_proc = train_dataset.size().value() / numranks;
+
+  //Generate dataloader
+  auto total_batch_size = 64;
+  auto batch_size_per_proc = total_batch_size / numranks; //effective batch size in each processor 
+  auto data_loader = torch::data::make_data_loader
+                     (std::move(train_dataset), data_sampler, batch_size_per_proc);
+
+  // setting manual seed
+  torch::manual_seed(0);
+
+  auto model = std::make_shared<Model>();
+
+  auto learning_rate = 1e-2;
+
+  torch::optim::SGD optimizer(model->parameters(), learning_rate);
+
+  // Number of epochs
+  size_t num_epochs = 10;
+
+  // start timer
+  tstart = MPI_Wtime();
+
+  for (size_t epoch = 1; epoch <= num_epochs; ++epoch)
+  {
+      size_t num_correct = 0;
+
+      for (auto& batch : *data_loader)
+      {
+          auto ip = batch.data;
+          auto op = batch.target.squeeze();
+
+          //convert to required formats
+          ip = ip.to(torch::kF32);
+          op = op.to(torch::kLong);
+
+          //Reset gradients
+          model->zero_grad();
+
+          //Execute forward pass
+          auto prediction = model->forward(ip);
+
+          auto loss = torch::nll_loss(torch::log_softmax(prediction, 1) , op);
+
+          /*
+          //Print loss
+          if (rank == 0) {
+              std::cout << "Loss in epoch " << epoch << " = " << loss.item<float>() << std::endl;
+          }
+          */
+
+          //Backpropagation
+          loss.backward();
+
+          //Averaging the gradients of the parameters in all the processors
+          for (auto &param : model->named_parameters())
+          {
+              MPI_Allreduce(MPI_IN_PLACE, param.value().grad().data_ptr(),
```

Comment:
cc @agolynski, have we exposed ProcessGroup C++ APIs? Can users call allreduce directly from process group instances?